Update modbus test harness

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The test harness mocked modbus.setup() and thus disabled testing of a the base modbus, in order to test components. This worked fine (apart from the missing test) with the “old” style configuration, but failed completely for the “new” style (discovery_info).

The new test harness mocks class modbusHub, which is the class isolating the pymodbus lib, therefore all code in the integration are tested.

The test harness itself was simplified based on the experience from adding component tests. Writing test cases are now a real simple case, and once this PR is merged, test cases will be demanded for every PR.

In order to do more live testing PR42120 together with the test cases have been used. Changes to PR42120 are available in https://github.com/janiversen/core/tree/PR42120_update

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
NO change
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ x] The code change is tested and works locally.
- [ x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ x] There is no commented out code in this PR.
- [ x] I have followed the [development checklist][dev-checklist]
- [ x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
Update test harness to allow discovery_info testing.

This is a needed update, before the whole config is converted
to the new structure, because it allows test of both the old
and new configuration style.

The following changes have been made:
- Combine test functions into a single base_test.
- Prepare for new mock by combining the 2 common test functions into one.
- Change mock to be only modbusHub
- Do not replace whole modbus class, but only the class that
  connects to pymodbus (modbusHub). This allows test of modbus
  configurations and not only component configurations, and is needed
  when testing discovery activated platform.
- Add test of climate.
  Warning this is merely test of discovery,
  the real cover tests still needs to be added.
- Add test of cover.
  Warning this is merely test of discovery,
  the real cover tests still needs to be added.

janiversen commented:
@vzahradnik please do a review, when you have time. Remark the branch I have prepared for your PR 42120

janiversen commented:
How nice to be able to test with both old and new configurations, it saves me a lot of time while "upgrading" the whole configuration to the new way.

janiversen commented:
Found a small error, will update the PR in a couple of hours.

janiversen commented:
and solved. Ready for review.

@vzahradnik I have also update my branch PR42120_update, because there was a couple of problems in light.py.

I am nearly completed with the configuration update, but it depends on this PR, so I hope it will be reviewed soon and merged.

vzahradnik commented on tests/components/modbus/test_modbus_climate.py:
```diff
@@ -0,0 +1,47 @@
+"""The tests for the Modbus cover component."""
```
Please fix the typo. Not cover, but climate component.

janiversen replied:
Guilty as charged, that happens when you copy files :-) I have fixed it in an additional commit.

vzahradnik reviewed (changes requested):
@janiversen great work! Apart from finding one typo in the comment, your changes look good to me.

janiversen commented:
Combined all the commits into one, to make life easier for the maintainer that hopefully one day will merge the PR.

@vzahradnik there are no changes in this squash, and I corrected the typo, so could you please approve the PR.

janiversen commented:
@MartinHjelmare 2 questions:
1) any chance you could look at this soon (it us ready from our side), and it holds up our conversion to the  new configuration (thanks in advance).
2) I think there is a problem with the new way of making configurations. All devices needs to be within the “modbus:” and “-platform” are part if the old way.

How is this supposed to work for integrations that use modbus? The platform statement gave us a nice way of depending on another integration, without mixing the code.

as one of the modbus codeowners I help and encourage other integrations to use modbus, and not repite the code, but I would not like to see all these integrations in the modbus directory.

Please let me know what your suggestion is?

One way could be to have subdirectories in modbus for each “foreign” integration with a new CODEOWNER etc.

MartinHjelmare commented on tests/components/modbus/conftest.py:
```diff
@@ -37,61 +40,84 @@ def __init__(self, register_words):
-        assert await async_setup_component(hass, entity_domain, config)
+        # setup modbus platform
+        if method_discovery:
+            await async_load_platform(
```
Setting up the integration should also set up the platforms. Why do we need to do that in the test specifically?

MartinHjelmare commented on tests/components/modbus/conftest.py:
```diff
@@ -37,61 +40,84 @@ def __init__(self, register_words):
 
+    # Setup inputs for the sensor
+    read_result = ReadResult(register_words)
+    hub: ModbusHub = hass.data[DOMAIN][DEFAULT_HUB]
```
We shouldn't interact with details in the tests. Patch the library as needed to avoid I/O and use the core interfaces to start the tests.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

janiversen replied:
I see your point, however if I use Hass.states to set the state, the whole idea of testing if void, because what we want to test is that the result from the pymodbus library is correctly converted to a state.

hub: ModbusHub = hass.data[DOMAIN][DEFAULT_HUB] retrieves a MagicMock and read_result is used as return value in a function in the MagicMock, so MagicMock replaces the pymodbus library (actually the class we added on top of the library).

Due to MagicMock there are no IO not is pymodbus ever called in test.

If I use Hass.states to set state, there are no way to test the conversion from "raw" input to a well defined state.
(remark this is not new code, this have been in here since the beginning).

MartinHjelmare replied:
We don't want to set fake states with hass.states.

Set up the integration but don't interact with any integration details, like accessing hass.data. `hass.data` holds internal objects that we shouldn't rely on in our tests.

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -71,9 +66,8 @@
-    SERVICE_WRITE_COIL,
-    SERVICE_WRITE_REGISTER,
 )
+from .modbus import modbus_setup
```
Is this change intended for this PR? We should preferably only refactor the tests and not the code under test at the same time if possible.

janiversen commented:
I was not aware that I could use manifest.json to add a dependency of another integration, it seems only to have third party libraries. I made a short test, and saw pip trying to install the other integrations, but I might have done something wrong.

However using manifest.json, would mean the new integration would have to mimic the whole modbus.setup as well as the SCHEMA declarations, that can hardly be seen as user friendly.

today you have “modbus:” with its configuration and the new intergration with
foo:
       -platform modbus

and we cannot have
modbus:
     foos: []

without mixing the code.

Have I completely misunderstood something:

janiversen commented:
I will push an update in a moment because a new PR for modbus was merged independent of our request for changes.

janiversen commented:
This does not seem to be my problem:
ERROR tests/components/shelly/test_config_flow.py::test_form_errors_test_connection[pyloop-error1]
2486
ERROR tests/components/sharkiq/test_vacuum.py::test_device_properties[pyloop-name-Sharknado]

MartinHjelmare commented:
Here are the docs for `dependencies`:

https://developers.home-assistant.io/docs/creating_integration_manifest#dependencies

MartinHjelmare reviewed (commented):
My previous questions and comments haven't been addressed as far as I can see. Did I miss it?

janiversen committed:
Update comment for old config

janiversen committed:
Do not use hass.data, but instead patch pymodbus library.

janiversen reviewed (commented):
Thanks a lot, that is really a nice way to interlink integrations.

I have adressed all your comments:
- code / test are split (you already merged my PR, thanks)
- Comments are made clearer why we load modbus + devices for the old config and only modbus for discovery_info.
- Hass.data usage.
I have to thanks you for the Hass.data comment, I was lazy. It made me look for a better solution, and I found it. I am patching the pymodbus library so ALL code in the modbus integration is now testable !!

Please let me know if you have further comments or if it can be merged.

janiversen committed:
Add test of configuration (old/new way as available).

janiversen commented:
While waiting, I back ported the test of configurations, this really belongs in this PR, and makes the next PR a lot cleaner.

MartinHjelmare commented on tests/components/modbus/conftest.py:
```diff
@@ -37,63 +32,101 @@ def __init__(self, register_words):
+                        {array_name_discovery: [{**config_device}]}
+                    )
+                await async_load_platform(
+                    hass, DOMAIN, DOMAIN, config_modbus[DOMAIN], config_modbus
```
There's no modbus platform. The platforms are cover, sensor, switch etc in the modbus integration.

For the platforms that have been migrated to be loaded via discovery, like cover, it's enough to set up the modbus integration with `async_setup_component`.

janiversen replied:
you got me there :-) First I understood the difference between "integration" and "platform"; but then I save the calls "setup_component" which seems to indicate something in between.

Anyhow I have corrected it, and thanks for explaining the difference.

MartinHjelmare replied:
We used to call integrations "components".

MartinHjelmare commented on tests/components/modbus/conftest.py:
```diff
@@ -37,63 +32,101 @@ def __init__(self, register_words):
+
+        # Check state
+        entity_id = f"{entity_domain}.{device_name}"
+        state = hass.states.get(entity_id).state
+        assert state == expected
```
It's more readable if we keep the assertion in the tests and just have the setup code in this helper function.

janiversen replied:
Agreed and done. On top of that I have added a cover function base_config_test, so it is very clear in the test functions which functionality is wanted (test or just configuration check).

Hope that solves your comments.

janiversen committed:
Move assert to test function.

Make assert more understandable by removing it from the helper.

add base_config_test (check just calls base_test) to make it clear if
test is wanted or just controlling the config.

janiversen committed:
use setup_component to load Modbus since it is an integration.

janiversen committed:
Optimized flow in test helper.

MartinHjelmare reviewed (approved):
Looks good!
